crates/dojo-lang/src/model.rs

```diff
@@ -236,81 +238,153 @@ pub fn handle_model_struct(
-            diagnostics.push(PluginDiagnostic {
-                message: "Key is only supported for core types that are 1 felt long once \
-                          serialized. `u256` is a struct of 2 u128, hence not supported."
-                    .into(),
-                stable_ptr: struct_ast.name(db).stable_ptr().untyped(),
-                severity: Severity::Error,
-            });
-        }
-    }
-
-    let serialize_member = |m: &Member, include_key: bool| {
-        if m.key && !include_key {
-            return None;
-        }
-
-        if m.ty == "felt252" {
-            return Some(RewriteNode::Text(format!(
-                "core::array::ArrayTrait::append(ref serialized, *self.{});",
-                m.name
-            )));
-        }
-
-        Some(RewriteNode::Text(format!(
-            "core::serde::Serde::serialize(self.{}, ref serialized);",
-            m.name
-        )))
-    };
-
-    let serialized_keys: Vec<_> =
-        keys.iter().filter_map(|m| serialize_member(m, true)).collect::<_>();
-
-    let serialized_values: Vec<_> =
-        members.iter().filter_map(|m| serialize_member(m, false)).collect::<_>();
-
     let name = struct_ast.name(db).text(db);
     aux_data.models.push(Model {
         name: name.to_string(),
         namespace: model_namespace.clone(),
-        members: members.to_vec(),
+        members,
     });
 
     (
         RewriteNode::interpolate_patched(
             "
-impl $type_name$Model of dojo::model::Model<$type_name$> {
+#[derive(Drop, Serde)]
+pub struct $type_name$Values {
+    $members_values$
+}
+#[generate_trait]
+impl $type_name$Model of $type_name$Trait {
+    fn entity_id_from_keys($param_keys$) -> felt252 {
+        let mut serialized = core::array::ArrayTrait::new();
+        $serialized_param_keys$
+        core::poseidon::poseidon_hash_span(serialized.span())
+    }
+    fn get(world: dojo::world::IWorldDispatcher, $param_keys$) -> $type_name$ {
+        let mut serialized = core::array::ArrayTrait::new();
+        $serialized_param_keys$
+        dojo::model::Model::<$type_name$>::entity(
+            world,
+            serialized.span(),
+            dojo::model::Model::<$type_name$>::layout()
+        )
+    }
+    fn set(self: @$type_name$, world: dojo::world::IWorldDispatcher) {
+           dojo::model::Model::<$type_name$>::set_entity(
+            world,
+            dojo::model::Model::<$type_name$>::keys(self),
+            dojo::model::Model::<$type_name$>::values(self),
+            dojo::model::Model::<$type_name$>::layout()
+        )
+    }
+    fn delete(self: @$type_name$, world: dojo::world::IWorldDispatcher) {
+        dojo::model::Model::<$type_name$>::delete_entity(
+            world,
+            dojo::model::Model::<$type_name$>::keys(self),
+            dojo::model::Model::<$type_name$>::layout()
+        )
+    }
+
+    $field_accessors$
+}
+impl $type_name$ModelValues of dojo::model::ModelValues<$type_name$Values> {
```

Comment:
As we plan to have some configuration from the user which part of the code they want to generate, couldn't the `ModelValues` also benefit from the entity member `set/get`?

Reply:
Yes, good idea 👍